Sources/Private/Experimental/Layers/AnimationLayer.swift

```diff
@@ -17,36 +17,67 @@ protocol AnimationLayer: CALayer {
 
 // Context describing the timing parameters of the current animation
 struct LayerAnimationContext {
+  /// The animation being played
+  let animation: Animation
+
   /// The timing configuration that should be applied to `CAAnimation`s
-  let timingConfiguration: ExperimentalAnimationLayer.TimingConfiguration
+  let timingConfiguration: ExperimentalAnimationLayer.CAMediaTimingConfiguration
 
   /// The absolute frame number that this animation begins at
   let startFrame: AnimationFrameTime
 
   /// The absolute frame number that this animation ends at
   let endFrame: AnimationFrameTime
-
-  /// The frame rate that this animation is played at
-  let framerate: CGFloat
-
-  /// The duration of this animation, in seconds
-  var duration: TimeInterval {
-    let frameDuration = endFrame - startFrame
-    return TimeInterval(frameDuration / framerate)
-  }
 }
 
 // MARK: - CAAnimation + LayerAnimationContext
 
 extension CAAnimation {
-  /// Configures the timing properties of this `CAAnimation`
-  /// using the current `LayerAnimationContext`
-  func configureTiming(with context: LayerAnimationContext) {
-    duration = context.duration
-    repeatCount = context.timingConfiguration.repeatCount
-    autoreverses = context.timingConfiguration.autoreverses
-    timeOffset = context.timingConfiguration.timeOffset
-    isRemovedOnCompletion = false
-    fillMode = .both
+  /// Creates a `CAAnimation` that wraps this animation,
+  /// applying timing-related configuration from the given `LayerAnimationContext`
+  func timed(with context: LayerAnimationContext) -> CAAnimation {
+
+    // The base animation always has the duration of the full animation,
+    // since that's the time space where keyframing and interpolating happens.
+    // So we start with a simple animation timeline from 0% to 100%:
+    //
+    //  ┌──────────────────────────────────┐
+    //  │           baseAnimation          │
+    //  └──────────────────────────────────┘
+    //  0%                                100%
+    //
+    let baseAnimation = self
+    baseAnimation.duration = context.animation.duration
+    baseAnimation.speed = (context.endFrame < context.startFrame) ? -1 : 1
+
+    // To select the subrange of the `baseAnimation` that should be played,
+    // we create a parent animation with the duration of that subrange
+    // to clip the `baseAnimation`. This parent animation can then loop
+    // and/or autoreverse over the clipped subrange.
+    //
+    //        ┌────────────────────┬───────►
+    //        │   clippingParent   │  ...
+    //        └────────────────────┴───────►
+    //       25%                  75%
+    //  ┌──────────────────────────────────┐
+    //  │           baseAnimation          │
+    //  └──────────────────────────────────┘
+    //  0%                                100%
+    //
+    let clippingParent = CAAnimationGroup()
+    clippingParent.animations = [baseAnimation]
```

Comment:
Do we need this clipping parent in the scenario where the full animation is playing? Or could we return `baseAnimation` above if the clipping parent isn't needed?

Reply:
We don't strictly need the parent `CAAnimationGroup` when playing the full animation, but I think having just one codepath makes this setup simpler and easier to reason about. (I don't expect there to be an observable perf difference between the two, since this already has 0 CPU overhead).